core/client/src/main/java/alluxio/client/block/stream/AbstractPacketReader.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block.stream;
+
+import alluxio.Configuration;
+import alluxio.Constants;
+import alluxio.PropertyKey;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import io.netty.buffer.ByteBuf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * An abstract packet reader implementation with the actual reading part (the producer)
+ * unimplemented.
+*/
+@NotThreadSafe
+public abstract class AbstractPacketReader implements PacketReader {
```

Comment:
I dislike this structure as well. I did this way because I was planning to implement LocalFilePacketReader via AbstractPacketReader so that the file reading and client calls to "read" can be parallelized like what NettyPacketReader does. Not sure whether that will happen. Given that you have this concern, I am fine with removing it.